engine/src/state.rs

```diff
@@ -0,0 +1,171 @@
+        fn as_ref(&self) -> &[u8] {
+            match self {
+                Self::NotFound => ERR_STATE_NOT_FOUND,
+                Self::SerializationFailed => ERR_STATE_SERIALIZATION_FAILED,
+                Self::DeserializationFailed => ERR_STATE_CORRUPTED,
+            }
+        }
+    }
+}
+
+pub mod legacy {
+    use super::*;
+
+    /// Migrates the state from the old format to the latest format.
+    ///
+    /// While this function works for a single v1 -> v2 migration, it would need to be
+    /// altered to handle all earlier migrations if needed.
+    pub fn migrate_state<I: IO>(io: &mut I) -> Result<(), error::EngineStateError> {
+        // Get the engine legacy state.
+        let engine_state_legacy = match io.read_storage(&bytes_to_key(KeyPrefix::Config, STATE_KEY))
+        {
+            None => Err(error::EngineStateError::NotFound),
+            Some(bytes) => EngineStateLegacy::try_from_slice(&bytes.to_vec())
+                .map_err(|_| error::EngineStateError::DeserializationFailed),
+        }?;
+
+        // Migrate to new engine state.
+        let engine_state_new = EngineState {
+            chain_id: engine_state_legacy.chain_id,
+            owner_id: engine_state_legacy.owner_id,
+            upgrade_delay_blocks: 0,
+            default_gas_token: Default::default(),
+        };
+
+        // Set the new engine state.
+        io.write_storage(
+            &bytes_to_key(KeyPrefix::Config, STATE_KEY),
+            &engine_state_new
+                .try_to_vec()
+                .map_err(|_| error::EngineStateError::SerializationFailed)?,
+        );
+
+        Ok(())
+    }
+
+    /// Engine internal state V1, mostly configuration.
+    /// Should not contain anything large or enumerable.
+    #[derive(BorshSerialize, BorshDeserialize, Default, Clone, PartialEq, Eq, Debug)]
+    pub struct EngineStateLegacy {
+        /// Chain id, according to the EIP-155 / ethereum-lists spec.
+        pub chain_id: [u8; 32],
+        /// Account which can upgrade this contract.
+        /// Use empty to disable updatability.
+        pub owner_id: AccountId,
+        /// Account of the bridge prover.
+        /// Use empty to not use base token as bridged asset.
+        pub bridge_prover_id: AccountId,
+        /// How many blocks after staging upgrade can deploy it.
+        pub upgrade_delay_blocks: u64,
+    }
+}
+
+#[cfg(test)]
+#[cfg(feature = "std")]
+mod tests {
+    use super::*;
+    use aurora_engine_test_doubles::io::{Storage, StoragePointer};
+    use std::sync::RwLock;
+
+    #[test]
+    fn test_missing_engine_state_is_not_found() {
+        let storage = Storage::default();
+        let storage = RwLock::new(storage);
+        let io = StoragePointer(&storage);
+
+        let actual_error = get_state(&io).unwrap_err();
+        let actual_error = std::str::from_utf8(actual_error.as_ref()).unwrap();
+        let expected_error = std::str::from_utf8(error::ERR_STATE_NOT_FOUND).unwrap();
```

Comment:
It seems that creating utf8 strings is redundant. You can just compare `&[u8]` instead.
```rust
assert_eq!(actual_error.as_ref(), error::ERR_STATE_NOT_FOUND);
```

Reply:
Will rework it. Part of code that I refactored, but haven't looked into improving yet. I also don't necessarily like the `RwLock`, seems to be a bit overkill.